File: R/utils/log_add.R
log_add = function(log_a, log_b) {
  # evaluate log(c) = log(a + b) when given log(a) and log(b)
  if(all(is.infinite(log_a), is.infinite(log_b))) {
    return(-Inf)
  }
  x = log_a - log_b
  exp_x = exp(x)
  if(is.infinite(exp_x)) {
    warning('exp(log(a) - log(b)) is infinite')
    return(log_a)
  }
  if(exp_x == 0) {
    warning('exp(log(a) - log(b)) is 0')
    return(log_b)
  }
  return(log_b + log(1 + exp_x))
}

log_sum = function(x) {
  # evaluate log(c) = log(x1 + ... + xn) when given log(x1), ..., log(xn)
  
  log_res = x[1]
  len = length(x)
  if(len > 1) {
    for(i in 2:len) {
      log_res = log_add(log_res, x[i])
    }
  }
  log_res
}

File: packages/dsmovetools2d/src/log_add.cpp
//
// Created by Joshua Hewitt on 3/15/21.
//

#include "Rcpp.h"

/*
 * implement log(c) = log(a + b) given log(a) = v, log(b) = data[i],
 * and log(c) = data[i] (updated).  uses the identity:
 *   log(c) = log(a + b) = log(b) + log( 1 + exp(log(a) - log(b)) )
 *
 * corrections are made to account for the possibility that b = 0
 */
double log_add(double log_a, double log_b) {
    if(!std::isfinite(log_a)) {
        if(log_a < 0) {
            return log_b;
        }
    }
    if(!std::isfinite(log_b)) {
        if(log_b < 0) {
            return log_a;
        }
    }
    double x = log_a - log_b;
    double exp_x = exp(x);
    // evaluate log(c)
    if(exp_x == HUGE_VAL) { // exp_x == "Inf"
        return log_a;
    }
    if(exp_x == 0) { // a has negligible size relative to b
        return log_b;
    }
    return log_b + std::log1p(exp_x);
}